src/libopenrave/kinbody.cpp

```diff
@@ -3467,84 +3539,117 @@ void KinBody::_ComputeInternalInformation()
                 }
                 else {
-                    dofformat.dofindex = (*itjoint)->GetDOFIndex();
-                    dofformat.jointindex = (*itjoint)->GetJointIndex();
+                    thisdofformat.dofindex   = pjoint->GetDOFIndex();  ///< >= 0
+                    thisdofformat.jointindex = pjoint->GetJointIndex(); ///< in [0, nActiveJoints)
                 }
-                for(int idof = 0; idof < (*itjoint)->GetDOF(); ++idof) {
-                    dofformat.axis = idof;
-                    if( !!(*itjoint)->_vmimic[idof] ) {
-                        // only add if depends on mimic joints
-                        FOREACH(itdofformat,(*itjoint)->_vmimic[idof]->_vdofformat) {
-                            JointPtr pjoint = itdofformat->GetJoint(*this);
-                            if( pjoint->IsMimic(itdofformat->axis) ) {
-                                mapmimic[dofformat] = (*itjoint)->_vmimic[idof];
+
+                const int ndof = pjoint->GetDOF();
+                const boost::array<MimicPtr, 3>& vmimic = pjoint->_vmimic;
+                for(int idof = 0; idof < ndof; ++idof) {
+                    const MimicPtr& pmimic = vmimic[idof]; // enumerate
+                    thisdofformat.axis = idof;
+                    if( !!pmimic ) {
+                        // only add if pjoint depends on mimic joints
+                        // TGN: Can an active joint depend on mimic joints??? If not, why need vjoints = _vecjoints?
+                        for(const Mimic::DOFFormat& dofformat : pmimic->_vdofformat) {
+                            const JointPtr pjointDepended = dofformat.GetJoint(*this);
+                            if( pjointDepended->IsMimic(dofformat.axis) ) {
+                                mapmimic[thisdofformat] = pmimic; ///< pjoint depends on pjointDepended
+                                RAVELOG_DEBUG_FORMAT("mimic joint %s depends on mimic joint %s; thisdofformat: %s; pmimic: %s",
+                                                     pjoint->GetName() % pjointDepended->GetName() % thisdofformat.to_string() % pmimic->to_string()
+                                );
                                 break;
                             }
                         }
                     }
                 }
-                ++jointindex;
             }
         }
+
         bool bchanged = true;
         while(bchanged) {
             bchanged = false;
-            FOREACH(itmimic,mapmimic) {
-                boost::shared_ptr<Mimic> mimic = itmimic->second;
-                Mimic::DOFHierarchy h;
-                h.dofformatindex = 0;
-                FOREACH(itdofformat,mimic->_vdofformat) {
-                    if( mapmimic.find(*itdofformat) == mapmimic.end() ) {
-                        continue; // this is normal, just means that the parent is a regular dof
+            for(const std::pair<const Mimic::DOFFormat, MimicPtr>& keyvalue : mapmimic) {
```

Comment:
@tgn3000 i'm not sure openrave users are using a compiler with this fancy feature enabled...

Reply:
This is just c++11 for-loop. In `CMakeLists.txt` we already had
```
set(CMAKE_CXX_STANDARD 11)
```